### ISSUE TITLE
lsp inlay hints

### ISSUE THREAD

**Issue by MangoIV** [mrcjkb/haskell-tools.nvim#364]
### Feature description

It would be super cool if you could support the new [lsp inlay hints](https://gpanders.com/blog/whats-new-in-neovim-0.10/#lsp-inlay-hints) for code lenses, they look much better than what we have now. I also wonder if there's a possibility to make code lenses look a tad nicer, e.g. not include control sequences and excessive amounts of white space (especially in imports). 

This will start working with neovim version 0.10! 

Thanks for considering!

**Comment by MangoIV** [comment]
Ah I think this is something to be supported by the hls? What kinda thing are the virtual text then? Why are they not supported by neovim proper? Hmmmm

**Comment by mrcjkb** [comment]
Hey :smile: 

inlay hints and code lenses are different LSP requests.
It's on the hls roadmap, but I have no idea when it will be done https://github.com/haskell/haskell-language-server/issues/2938.

Activity seems to have picked up recently though.

**Comment by MangoIV** [comment]
https://github.com/haskell/haskell-language-server/pull/4131

There’s a GSOC student working on this!

**Comment by MangoIV** [comment]
Do you know what’s up with the control sequences and the spaces on code lenses? Is that something that hls does wrong? Or is this something that happens because all of these guys are written for vscode?

**Comment by MangoIV** [comment]
do you just get these as strings?

**Comment by mrcjkb** [comment]
haskell-tools doesn't do much, besides enable and auto refresh code lenses by default.
The feature itself is [implemented in neovim](https://github.com/neovim/neovim/blob/4b029163345333a2c6975cd0dace6613b036ae47/runtime/lua/vim/lsp/codelens.lua?plain=1).

**Comment by MangoIV** [comment]
interesting, why is it that the standard lsp client doesn’t display these then? Also do you know where I’d have to go with these issues? I e what may cause the visual noise displayed in some of the code lenses?

**Comment by mrcjkb** [comment]
> why is it that the standard lsp client doesn’t display these then?

I think that's just a decision by the maintainer. It could add quite a bit of noise and be too controversial to be enabled by default in neovim core.
@mfussenegger, maybe you can answer this better than I can? :smile: 

> I e what may cause the visual noise displayed in some of the code lenses?

I'm not sure there's a good solution to this, because it would involve overriding `vim.lsp.codelens.on_codelens`.
I [do this in my personal config](https://github.com/mrcjkb/nvim/blob/0aa1fad6ddf7cb1c6d9109811bf56dd0f07eafeb/lib/lua/mrcjk/lsp/init.lua?plain=1#L34), to add a prefix to the title, but I never really noticed any noise - Do you have a concrete example/screen shot?).
That would be too hacky for haskelll-tools though.

**Comment by MangoIV** [comment]
<img width="1055" alt="IMG_3385" src="https://github.com/mrcjkb/haskell-tools.nvim/assets/40720523/140df788-1591-414c-afe2-51a3fb425ff4">

this is an example. It has weird amount of spaces, no?

**Comment by MangoIV** [comment]
There’s also the case where it renders “line breaks” in inferred type signatures.

**Comment by MangoIV** [comment]
![image](https://github.com/mrcjkb/haskell-tools.nvim/assets/40720523/17704e57-4358-44a4-ac5f-2dddf83ee69d)

**Comment by MangoIV** [comment]
as you can see it inserts these `^@` and then again, a weird amount of spaces :D

**Comment by MangoIV** [comment]
![image](https://github.com/mrcjkb/haskell-tools.nvim/assets/40720523/1c353420-7495-4031-9eeb-15351e2a990c)

**Comment by MangoIV** [comment]
I think so, yes, although it has also reliable been messing up these for me 😂

**Comment by MangoIV** [comment]
perhaps one could do something like replacing all consecutive whitespaces by one space in the code lenses? I think that would make it look a lot nicer...

**Comment by mrcjkb** [comment]
> perhaps one could do something like replacing all consecutive whitespaces by one space in the code lenses? I think that would make it look a lot nicer...

Perhaps. But I think this would have to be implemented in neovim core. Overriding the core function is a dirty hack, and would be brittle.

**Comment by MangoIV** [comment]
Yeah, now I only need to find how, thank you for dropping the link :)

**Comment by MangoIV** [comment]
https://github.com/neovim/neovim/pull/28807

I went off and fixed it 😅

**Comment by MangoIV** [comment]
```nix
package = pkgs.neovim-unwrapped.overrideAttrs (old: {
      patches = old.patches or [] ++ [
        (pkgs.fetchpatch {
          name = "strip-consecutive-white-spaces.patch"; 
          url = "https://github.com/mangoiv/neovim/commit/ea0a9163cb37e8444465b81abbed82c54134c12e.patch"; 
          hash = "sha256-LQwS+yJucin5AlXTPa3O95Zwjw4RerniPtNT8ctw0AU=";
        })
      ];
    });
```
